### PR TITLE
Token balances for erc20 tokens and eth

### PR DESCRIPTION
**Pull request by victorwiebe**

### CODE REVIEW

**Comment by VictorVicente** [comment]
You are naming every balance variable as `polyTokenBalance` and they are not always POLY. It's confusing, please rename them as `erc20tokenBalance` when required.

**Comment by VictorVicente** [comment]
Also, you are always checking the caller address but sometimes it's wrong. 
You should only check balances when a dividend is being created and it's ok to check caller balance's there.
It isn't necessary to check it for pull and push dividends methods. So you don't need any refactoring for DividendCheckpoint.

**Comment by victorwiebe** [comment]
> You are naming every balance variable as `polyTokenBalance` and they are not always POLY. It's confusing, please rename them as `erc20tokenBalance` when required.

Right, I will fix those

>Also, you are always checking the caller address but sometimes it's wrong.
You should only check balances when a dividend is being created and it's ok to check caller balance's there.
It isn't necessary to check it for pull and push dividends methods. So you don't need any refactoring for DividendCheckpoint. 
>I don't like the checkpoints refactoring.
I'd like to keep pushDividendPaymentToAddresses, pushDividendPayment and pullDividendPayment on DividendCheckpointWrapper to avoid duplicated code.
The best approach here is to move the code to a protected method and keep public methods in DividendCheckpointWrapper but calling this new protected methods.
This way no changes are required for EthDividendCheckpointWrapper.
For ERC20DividendCheckpointWrapper you only need to check balances first and then call the inherited protected method.

1) Where is callerAddress wrong? 
2) Understood, I wasn't sure if the refactor was right. If I change the DividendCheckpoint code to protected, would this be right for the ERC20DividendCheckpoint?

```
public pushDividendPaymentToAddresses = async (params: PushDividendPaymentToAddressesParams) => {
  < Check that the balance is correct >
super.pushDividendPaymentToAddresses(params)
  };
```

**Comment by VictorVicente** [comment]
> 1. Where is callerAddress wrong?
> 2. Understood, I wasn't sure if the refactor was right. If I change the DividendCheckpoint code to protected, would this be right for the ERC20DividendCheckpoint?
> 
> ```
> public pushDividendPaymentToAddresses = async (params: PushDividendPaymentToAddressesParams) => {
>   < Check that the balance is correct >
> super.pushDividendPaymentToAddresses(params)
>   };
> ```

There is no need to change DividendCheckpoint now. You should only add the check balance assert to all create dividends methods.

**Comment by victorwiebe** [comment]
> There is no need to change DividendCheckpoint now. You should only add the check balance assert to all create dividends methods.

But we **do** need to change ERC20DividendCheckpoint for push/pulldividend. As far as I understand, because of this code in the corresponding solidity method:
``` 
function _payDividend(address _payee, Dividend storage _dividend, uint256 _dividendIndex) internal {
        (uint256 claim, uint256 withheld) = calculateDividend(_dividendIndex, _payee);
        _dividend.claimed[_payee] = true;
        _dividend.claimedAmount = claim.add(_dividend.claimedAmount);
        uint256 claimAfterWithheld = claim.sub(withheld);
        if (claimAfterWithheld > 0) {
            require(IERC20(dividendTokens[_dividendIndex]).transfer(_payee, claimAfterWithheld), "transfer failed");
        }
        if (withheld > 0) {
            _dividend.totalWithheld = _dividend.totalWithheld.add(withheld);
            _dividend.withheld[_payee] = withheld;
        }
        emit ERC20DividendClaimed(_payee, _dividendIndex, dividendTokens[_dividendIndex], claim, withheld);
    }
```

File: src/contract_wrappers/modules/sto/usd_tiered_sto_wrapper.ts [hunk 1]
```diff
@@ -964,6 +971,7 @@ export default class USDTieredSTOWrapper extends STOWrapper {
     from: string,
     investmentValue: BigNumber,
     fundRaiseType: FundRaiseType,
+    usdToken: string | undefined,
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/sto/usd_tiered_sto_wrapper.ts]:
```suggestion
    usdToken?: string,
```
Make it optional, so you don't need to send `undefined`

File: src/contract_wrappers/tokens/security_token_wrapper.ts [hunk 2]
```diff
@@ -994,6 +999,13 @@ export default class SecurityTokenWrapper extends ERC20TokenWrapper {
     await this.checkOnlyOwner(params.txData);
     await this.checkModuleCostBelowMaxCost(params.address, maxCost);
     await this.checkModuleStructAddressIsEmpty(params.address);
+    if(params.budget) {
```
Comment (VictorVicente) [inline src/contract_wrappers/tokens/security_token_wrapper.ts]: It is not ok. 
You should get module cost from module factory and compare its value with the balance. 
I guess we are getting this cost on `this.checkModuleCostBelowMaxCost` so maybe it's easier to add this requirement there.

File: src/contract_wrappers/modules/checkpoint/erc20_dividend_checkpoint_wrapper.ts [hunk 3]
```diff
@@ -334,5 +354,14 @@ export default class ERC20DividendCheckpointWrapper extends DividendCheckpointWr
     }
     assert.isNonZeroETHAddressHex('token', token);
     assert.assert(name.length > 0, 'The name can not be empty');
+    let txDataPackage;
+    if (txData) {
+      txDataPackage = txData;
+    }
```
[the remark this answers is not in the document]
Comment (victorwiebe) [reply src/contract_wrappers/modules/checkpoint/erc20_dividend_checkpoint_wrapper.ts]: Awesome I understand now

File: src/contract_wrappers/tokens/security_token_wrapper.ts (hunk 2 above)
Comment (victorwiebe) [reply src/contract_wrappers/tokens/security_token_wrapper.ts]: In this case i am going to forget about budget and make a change inside `checkModuleCostBelowMaxCost`

**Pull request merged**